go/ethadapter/blockprovider.go

```diff
@@ -0,0 +1,210 @@
+	logger    gethlog.Logger
+	streamCh  chan *types.Block
+
+	// the state of the block provider
+	currentHead   *types.Header // most recently sent block (reset to nil when a `StartStreaming...` method is called)
+	runningStatus *int32        // 0 = stopped, 1 = running
+}
+
+// StartStreamingFromHash will look up the hash block, find the appropriate height (LCA if there have been forks) and
+// then call StartStreamingFromHeight based on that
+func (e *EthBlockProvider) StartStreamingFromHash(latestHash gethcommon.Hash) (<-chan *types.Block, error) {
+	ancestorBlk, err := e.latestCanonAncestor(latestHash)
+	if err != nil {
+		return nil, err
+	}
+	return e.StartStreamingFromHeight(increment(ancestorBlk.Number()))
+}
+
+// StartStreamingFromHeight will (re)start streaming from the given height, closing out any existing stream channel and
+// returning the fresh channel - the next block will be the requested height
+func (e *EthBlockProvider) StartStreamingFromHeight(height *big.Int) (<-chan *types.Block, error) {
+	// block heights start at 1
+	if height.Cmp(one) < 0 {
+		height = one
+	}
+	e.streamCh = make(chan *types.Block)
+	if e.getStatus() == stopped {
+		// if the provider is stopped (or not yet started) then we kick off the streaming processes
+		e.setStatus(running)
+		go e.streamBlocks(height)
+	}
+	return e.streamCh, nil
+}
+
+func (e *EthBlockProvider) Stop() {
+	e.setStatus(stopped)
+}
+
+func (e *EthBlockProvider) IsLive(h gethcommon.Hash) bool {
+	l1Head, f := e.ethClient.FetchHeadBlock()
+	return f && h == l1Head.Hash()
+}
+
+// streamBlocks is the main loop. It should be run in a separate go routine. It will stream catch-up blocks from requested height until it
+// reaches the latest live block, then it will block until next live block arrives
+// It blocks when:
+// - publishing a block, it blocks on the outbound channel until the block is consumed
+// - awaiting a live block, when consumer is completely up-to-date it waits for a live block to arrive
+func (e *EthBlockProvider) streamBlocks(fromHeight *big.Int) {
+	for e.getStatus() != stopped {
+		// this will block if we're up-to-date with live blocks
+		block, err := e.fetchNextCanonicalBlock(fromHeight)
+		if err != nil {
+			e.logger.Error("unexpected error while preparing block to stream, will retry in 1 sec", log.ErrKey, err)
+			time.Sleep(time.Second)
+			continue
+		}
+		e.logger.Trace("blockProvider streaming block", "height", block.Number(), "hash", block.Hash())
+		e.streamCh <- block // we block here until consumer takes it
+		// update stream state
+		e.currentHead = block.Header()
+	}
+}
+
+func (e *EthBlockProvider) fetchNextCanonicalBlock(fromHeight *big.Int) (*types.Block, error) {
+	if e.currentHead == nil {
+		blk, err := e.ethClient.BlockByNumber(fromHeight)
+		if err == nil {
+			return blk, nil
+		}
+	}
+
+	l1Block, f := e.ethClient.FetchHeadBlock()
+	if !f {
+		return nil, errors.New("l1 head block not found")
+	}
+	l1Head := l1Block.Header()
```

Comment:
Do you have a suggestion for a better name? For me it feels pretty accurate, it's the head of the L1.

Tbh I think `latestSent` was a clearer name than `currentHead` because currentHead makes it sound like some sort of source of truth but really it's just the block we most recently sent to our consumer.

L1 client -> [block provider] -> ordered block consumer

`l1Head` is the head block of the L1 according to the L1 client
`currentHead`/`latestSent` is the block sent most recently to the consumer

Reply:
Adding to the confusion here is the ambiguous words like `head` in "head block" vs "header", `block` in "L1 block" vs the "function will block" haha